espresso/streamer.go

```diff
@@ -193,40 +193,125 @@ func (s *EspressoStreamer[B]) CheckBatch(ctx context.Context, batch B) (BatchVal
-	if err != nil {
-		return 0, 0, fmt.Errorf("failed to fetch HotShot block height: %w", err)
-	}
-	start := s.hotShotPos
-	finish := min(start+100, currentBlockHeight)
-
-	return start, finish, nil
+// HOTSHOT_BLOCK_LOAD_LIMIT is the maximum number of blocks to attempt to
+// load from Espresso in a single process. This helps to limit our block
+// polling to a limited number of blocks within a single batched attempt.
+const HOTSHOT_BLOCK_LOAD_LIMIT = 100
+
+// computeEspressoBlockHeightsRange computes the range of block heights to fetch
+// from Espresso. It starts from the last processed block and goes up to
+// HOTSHOT_BLOCK_LOAD_LIMIT blocks ahead or the current block height, whichever
+// is smaller.
+func (s *EspressoStreamer[B]) computeEspressoBlockHeightsRange(currentBlockHeight uint64) (start uint64, finish uint64) {
+	start = s.hotShotPos
+	finish = min(start+HOTSHOT_BLOCK_LOAD_LIMIT, currentBlockHeight)
+
+	return start, finish
 }
 
-// / Update the batch buffer by reading from the Espresso blocks
-// / @param ctx context
-// / @return error possible error
+// Update will update the `EspressoStreamer“ by attempting to ensure that the
+// next call to the `Next` method will return a `Batch`.
+//
+// It attempts to ensure the existence of a next batch, provided no errors
+// occur when communicating with HotShot, by processing Blocks retrieved from
+// `HotShot` in discreet batches. If each processing of a batch of blocks will
+// not yield a new `Batch`, then it will continue to process the next batch
+// of blocks from HotShot until it runs out of blocks to process.
+//
+//	NOTE: this method is best effort.  It is unable to guarantee that the
+//	next call to `Next` will return a batch.  However, the only things
+//	that will prevent the next call to `Next` from returning a batch is if
+//	there are no more HotShot blocks to process currently, or if an error
+//	occurs when communicating with HotShot.
 func (s *EspressoStreamer[B]) Update(ctx context.Context) error {
-	// Fetch more batches from HotShot if available.
-	start, finish, err := s.computeEspressoBlockHeightsRange(ctx)
+	currentBlockHeight, err := s.EspressoClient.FetchLatestBlockHeight(ctx)
 	if err != nil {
 		return err
 	}
 
+	// Fetch more batches from HotShot if available.
+	start, finish := s.computeEspressoBlockHeightsRange(currentBlockHeight)
 	s.Log.Info("Fetching hotshot blocks", "from", start, "upTo", finish)
+	i := start
 
+	for i <= currentBlockHeight {
+		s.Log.Info("Remaining list before", "Size", len(s.RemainingBatches))
+
+		// Process the remaining batches
+		s.processRemainingBatches(ctx)
+
+		s.Log.Info("Remaining list after", "Size", len(s.RemainingBatches))
+		// Process the new batches fetched from Espresso
+		if err := s.processHotShotRange(ctx, i, finish); err != nil {
+			return fmt.Errorf("failed to process hotshot range: %w", err)
+		}
+		// Our position has advanced to the block following the finish block
+		// we just passed.
+		i = finish + 1
```

Comment:
I have the feeling that in some cases we might end up skipping some espresso blocks.
Why not just increment `i` by 1 and have it range through start...finish?

Reply:
Oh I guess I am confusing this loop with what happens in `processHotShotRange`.

Reply:
Right, the range is being processed in the `processHotShotRange`.  Provided that no error is returned from calling that method, it will process the given `start` to `finish` **INCLUSIVELY**, which is why I'm setting `i` to `finish` + 1. It's a small optimization, the increment **doesn't** actually need to happen, but doing it will allow us to not download the `finish` block again.  This loop here is tracking the position, just to ensure that we don't exceed the `currentBlockHeight` we've already pulled.

Reply:
I understand now thanks.